### ISSUE TITLE
Exception thrown frmo the validator rather then returning an error.

### ISSUE THREAD

**Issue by ericbmerritt** [jonasschmidt/ex_json_schema#7]
We get an error when passing a string, when the schema expects a map. It throws the map error seen below rather than the expected validation error. 

```
iex(13)> validate("test1.json", "88888888") 
   ** (BadMapError) expected a map, got: "88888888"
   (stdlib) :maps.is_key("id", "88888888")
   (ex_json_schema) lib/ex_json_schema/validator.ex:116: anonymous fn/2 in     ExJsonSchema.Validator.validate_aspect/4
(elixir) lib/enum.ex:873: anonymous fn/3 in Enum.flat_map/2
(elixir) lib/enum.ex:1473: Enum."-reduce/3-lists^foldl/2-0-"/3
(elixir) lib/enum.ex:872: Enum.flat_map/2
(elixir) lib/enum.ex:873: anonymous fn/3 in Enum.flat_map/2
(stdlib) lists.erl:1262: :lists.foldl/3
(elixir) lib/enum.ex:872: Enum.flat_map/2
iex(13)> validate("test1.json", %{}) 
{:error, [{"Required property id was not present.", "#"}]}
```

test1.json is:
```
{
    "type": "object"
    , "$schema": "http://json-schema.org/draft-04/schema"
    , "id": "<obsfuscated>"
    , "properties": {
        "id": {
            "$ref": "/dir1/test2.json"
        }
        , "name": {
            "type": "string"
        }
    }
    , "required": ["id"]
}

```

test2.json is:
```
{
    "$schema": "http://json-schema.org/draft-04/schema"
    , "id": "https://api.sunlightpayments.com"
    , "type": "string"
    , "pattern": "^[0-9]{8}$"
}
```

**Comment by jonasschmidt** [comment]
The first argument has to be a `Map`, a filename won't work there. The reason is to keep the library free of dependencies like a JSON parser, and leave that choice to the user. If you're dealing with JSON payloads that you want to validate, you are bound to have a JSON parser around anyway.

Please have a look at the README for some usage examples.

**Comment by ericbmerritt** [comment]
@jonasschmidt  You missed an important bit. I am not talking about json encoded strings at all. If I have a schema that expects a string (a string, not json) then things are fine. If I have a schema that expects a map and I pass it a string (because I am taking input from a user or some such) again not a json encoded string, then I get an exception rather than an error about not matching the schema.

**Comment by jonasschmidt** [comment]
Oh, ok. I misunderstood your question because in your code sample it looked like you were passing a string as the schema. Then it does indeed seem like there's a problem somewhere, I just can't see it yet. Your schema looks correct, and if I try validating a string against that schema, it gives me the correct result:
```elixir
iex(2)> schema = %{"type" => "object", "$schema" => "http://json-schema.org/draft-04/schema", "id" => "<obsfuscated>", "properties" => %{"id" => %{"type" => "string", "pattern" => "^[0-9]{8}$"}, "name" => %{"type" => "string"} }, "required" => ["id"] }
%{"$schema" => "http://json-schema.org/draft-04/schema",
  "id" => "<obsfuscated>",
  "properties" => %{"id" => %{"pattern" => "^[0-9]{8}$", "type" => "string"},
    "name" => %{"type" => "string"}}, "required" => ["id"], "type" => "object"}
iex(3)> ExJsonSchema.Validator.validate(schema, "foo")
{:error, [{"Type mismatch. Expected Object but got String.", "#"}]}
```

**Comment by jonasschmidt** [comment]
Actually, I just remembered that this issue was fixed in `v0.4.0` (original issue is here: https://github.com/jonasschmidt/ex_json_schema/issues/5). I was thinking in a wrong direction because of the misunderstanding with the schema. Just upgrade to `v0.4.0` and your problem should go away :)

**Comment by ericbmerritt** [comment]
Ok. excellent. Thanks.
